[GStreamer][MSE] Make setAppendState thread safe

charlie-ht opened the pull request:
https://bugs.webkit.org/show_bug.cgi?id=175540

setAppendState is indirectly called by other threads, despite
assertion that existed about it only running on the main
thread. One way this can happen which was causing problems was in
response to the sync-message::need-context glib signal. This was
invoking setAppendState from a non-main thread.

calvaris reviewed (commented):
As in upstream, I think it is ok but I prefer to wait for @eocanha to give the green light

eocanha commented:
See the answer in https://bugs.webkit.org/show_bug.cgi?id=175540

I still have to think about it, but maybe we can split the assert(isMainThread()) and apply it everywhere except in the KeyNeeded->whatever transitions. If we don't find a better way to do it (by signalling the main thread through the bus and transitioning there), for instance because the main thread is already blocked waiting for the key, then these should be the only allowed exception for transitioning from a non-main thread.

charlie-ht commented:
Expanding in upstream bug report.

charlie-ht committed:
[GStreamer][MSE] Make setAppendState thread safe
https://bugs.webkit.org/show_bug.cgi?id=175540

setAppendState is indirectly called by other threads, despite
assertion that existed about it only running on the main
thread.

charlie-ht commented:
So after some discussion with @eocanha , I learned that @calvaris prefers using the GStreamer bus in lieu of a function dispatcher. I followed a similar pattern for posting on the main thread via the bus in two places that needed it.

There's some duplicated code now in the blocks that check `isMainThread` and either call the function on the main thread or dispatch to the bus. That could be factored away, but I'd prefer a review of the approach before polishing it further.

The upstream patch can be ready to go, it's just quite different from this one (there's only one place that needs fixing up, rather than two)

[P.S I force updated this branch]
